Modify runtests to add option to specify a response file for additional xunit parameters and use named parameters

cshung opened the pull request:
Fixes #2287

I am opening this pull request to kick start the effort to fix #2287.

This pull request contains a prototype named argument parser. I believe it is robust and sufficient for the purpose. It also contains a document on my thought about how should we proceed to actually make the change.

Please take a look and suggest what shall we do next?

ViktorHofer commented on andrewau-notes.md:
```diff
@@ -0,0 +1,43 @@
+2. Make the arg parser backward compatible, if the first argument is not starting with '-', it is probably the old caller and just leave the old logic around, this is just maintainence overhead.
+3. Actually enumerate all the callers the best we could, and leave the rest (if any) to "if crash, we will fix it"
+
+(1) would be my preferred choice, if we have confidence that (3) is feasible, then do (3) to avoid changes to generator.
```
I know all the consumers of the script as I make changes to it frequently. CoreFx in two places, CoreClr in one place and soon Mono. I recommend to make a hard cut - option 3.

hoyosjs replied:
Option 3 leaves us with less technical debt on the long run. Having 2 positional parameters was bad enough already.
That I know of there is:

CoreFX:
- https://github.com/dotnet/corefx/blob/master/eng/sendtohelix.proj#L78-L79
- run-test.sh on the repo root

Coreclr:
- Scripts for corefx test running under tests/scripts. That being said, this change shouldn't affect these call sites and they probably shouldn't be updated. They rely on the old scripts packaged in 2018. Once we consume this logic those wrappers should be rewritten completely.

(Buildtools has some references to runtests but they shouldn't matter here anymore)

cshung replied:
Thanks, @ViktorHofer, and @hoyosjs.
I will execute on the plan that we will patch the call sites.

ViktorHofer commented:
The shell parsed argument approach looks good. For the cmd script I expect something easier with less code. Let me research a bit.

ViktorHofer commented on andrewau-notes.md:
```diff
@@ -0,0 +1,43 @@
+Just in case we wanted more bash programming that is not available to copy, [here](https://en.wikibooks.org/wiki/Bash_Shell_Scripting) is probably a fine reference.
+
+# Subproblem 2
+The usual approach, let's find something to copy from, [This](https://github.com/dotnet/arcade/blob/master/eng/common/CIBuild.cmd) one seems useless, almost all cmd file we have simply delegate to powershell. That makes me wonder, shall we do that too? It incurs the cost of launching powershell and writing a new script in powershell syntax, probably not worth it just for the command line parser.
```
You are right not trying to use powershell. We are testing on some Windows distros which don't have powershell installed. The Powershell approach that we took with arcade with the build scripts is OK as we don't build on any Windows variants that don't have Powershell integrated.

cshung replied:
Look like this thread can be resolved without need any work, resolving as is.

cshung commented:
> The shell parsed argument approach looks good. For the cmd script, I expect something easier with less code. Let me research a bit.

With your thumbs up on the shell part, I have done with the updating of the template of the shell script. Once I have your research about the cmd script side, I will probably take your result and engineer the same for Windows.

The cmd script is essentially the same as the shell script, the complexity is only apparent. The constraint of the scripting language itself play big part in it. An effort is spent to make sure both scripts are behaviorally identical with the shell script one, as you can note in my test baselines. The idea is that some people (me included) sometimes only test on Windows and wish for the best for cross-plat, behavioral differences can be a source of surprise (if not pain).

It might appear that we are doing duplicated work in setting `argparser_runtime_root` and then export it again as `RUNTIME_ROOT`. The idea is that I don't want the script to leave a polluted environment in case of a premature exit. It is much better to export only after we know there is nothing wrong.

Let's get back to our subproblem 4 - since we have the shell script updated, in principle, we could test it now. Do you know how, if not impossible, to test the change in the template?

jorive commented on andrew-lab/parse.cmd:
```diff
@@ -0,0 +1,60 @@
+REM Prevent affecting possible callers of the batch
+REM Without delayed expansion, !arg%argno%! used below won't work.
+
+set configuration_specified=0
```
Some suggestions when writing batch:
```diff
REM 1. Wrap setting environment variables within double quotes
- set Configuration=Release
+ set "Configuration=Release"

REM 2. Remove the potential surrounding quotes from the variables before comparing for equality
- if -%1- == -- (
+ if "%~1" == "" (

REM 3. When using a variable as a flag, you could undefine it first, then check if defined
- set "MyFlag=0"
REM Somewhere in the script you set it, the check
- if "%MyFlag%"=="1" (
+ set "MyFlag="
REM Somewhere in the script you set it, then check
+ if defined MyFlag (

REM 4. When using exit /b, be explicit on the return value
- exit /b
+ exit /b 0
```

More information can be found here: [Command Line arguments (Parameters)](https://ss64.com/nt/syntax-args.html)

cshung replied:
Feedback is taken into the batch script.

cshung commented:
Here are the list of items done in the latest iteration:

- [x] Addressing the feedback for the batch script (thank you @jorive )
- [x] Tested the generated script using the existing unit tests I had - the generated script produced exactly the same result as my prototype did.
- [x] Updated the windows template

And here are a couple of things I identified that needs work:
- [ ] I remember @hoyosjs mentioned some feedback on the C# code generator. The code was meant to be one-off, generate the script, and be done with it. If we think there is a need to generate more argument parsing scripts, we should probably keep the generate code somewhere (any suggestion where?) and make sure it is maintained.
- [ ] I remember @jorive mentioned that he ran into multiple instances where special characters in argument value screw things up, it would be nice to make sure the parser is solid with respect to those inputs.
- [ ] It would be nice to have an explicit CoreFX sign-off, please @ViktorHofer, @safern.

hoyosjs commented on andrew-lab/Program.cs:
```diff
@@ -0,0 +1,221 @@
+set ""OUTPUTENVIRONMENTVARIABLENAMEPLACEHOLDER_specified=""
+";
+            string template3 = @"
+:argparser
```
Can we combine the longhand and shorthand holders? They generate branching statements with the same exact code save for the condition. I think batch OR is weird, but I can imagine something like:
```
if shortname, (increase OUTPUTENVIRONMENTVARIABLENAMEPLACEHOLDER_specified)
if longname,  (increase OUTPUTENVIRONMENTVARIABLENAMEPLACEHOLDER_specified))
if OUTPUTENVIRONMENTVARIABLENAMEPLACEHOLDER_specified greater than 1 error
if value empty, error
set value
```

cshung replied:
We investigated it briefly, we didn't find an obvious way to do arithmetic in batch script, so resolving this as won't fix.

jorive replied:
What kind of arithmetic do you need to do?

hoyosjs replied:
Simple accumulator (x = x+1) upon seeing one of the options

hoyosjs replied:
Simple accumulator (x = x+1) upon seeing one of the options

hoyosjs replied:
Simple accumulator (x = x+1) upon seeing one of the options

jorive replied:
Like this?
```cmd
set /A "COUNT=1"
for /l %%i in (1,1,10) do (
    set /A "COUNT+=%COUNT%"
)
echo %COUNT%
```

hoyosjs replied:
Indeed

jorive replied:
Batch Kung Fu!

cshung replied:
Thanks - the arithmetic trick is incorporated in the new version.

hoyosjs commented on src/Microsoft.DotNet.CoreFxTesting/build/assets/RunnerTemplate.Windows.txt:
```diff
@@ -1,24 +1,125 @@
-)
+set "argparser_runtime_path_specified="
 
-set EXECUTION_DIR=%~dp0
```
Please leave this here. With the help of @tommcdon we found that the shift is messing with your set later.

jorive replied:
Why do we need to capture `%~dp0` into a variable? Can we just use `%~dp0`?

cshung replied:
The bug @hoyosjs and @tommdon found is that shift changes the value of %~dp0 implicitly.
To capture the name of the execution directory, one must do that up front, otherwise the value of %~dp0 becomes something else after shift is executed.

The script template is updated so that it stays at the beginning of the script.

jorive replied:
makes sense :)

jorive replied:
You do not need to cache the `%~dp0` if you follow the pattern below (arguably cleaner):
```cmd
@if not defined _echo echo off
cls

call :argparse
call :argparse
call :argparse
call :argparse
exit /b 0

REM The shift happens inside the function. Once you are out, then `%~dp0` is intact.
:argparse
setlocal
  echo "dp0=%~dp0"
  shift
  shift
endlocal& exit /b 0
```

cshung commented on andrew-lab/Program.cs:
```diff
@@ -0,0 +1,221 @@
+
+        private static string GenerateBashScript(Argument[] arguments)
+        {
+            string template1 = @"
```
Give meaningful names to these items,
Make them class statics

cshung replied:
Fixed, those strings are now class static with meaningful names.

cshung commented:
In this iteration I introduced two commits - one commit addressed the feedback from @hoyosjs by both fixing the `%~dp0` bug and addressed the comments on the C# code generator. The C# code generator change is mostly cosmetic - moving local string variables to global static constant, changing the upper case `IF` to lower case `if` for consistency, and removed bogus new lines. The new script generator is subjected to the same unit test bed so they should be solid (with respect to those cases at least)

In the second iteration, I introduced the new `--xunit-rsp` argument. The argument addresses the issue that we are trying to fix. As of the change, the argument is stored in an environment variable and is not used anywhere. 

I'd still love to see more or less the same three items:
- [ ] Can we please suggest a proper home (other and `andrew-lab`) to store to argument parsing script generator code?
- [ ] I remember @jorive mentioned that he ran into multiple instances where special characters in argument value screw things up, it would be nice to make sure the parser is solid with respect to those inputs.
- [ ] It would be nice to have an explicit CoreFX sign-off, please @ViktorHofer, @safern.

My rotation for the infrastructure work is scheduled to end on April 15th - would love to see this pull request got merged by the date so that it can be consumed by @mikem8361 when he is back.

hoyosjs commented:
You got it :)

ViktorHofer commented:
I'll take a look later today.

jorive commented:
Basic Batch Arithmetic Example
```cmd
@if not defined _echo echo off

rem Need delayed expansion
setlocal enabledelayedexpansion

    rem Need to declare counter with /a
    set /a "count=0"

    for /l %%i in (1,1,10) do (
        rem Inside common blocks like this, need to use ! instead of %
        set /a "count=!count!+1"
    )
    echo "count=%count%"

endlocal& exit /b 0
```

hoyosjs commented on src/Microsoft.DotNet.CoreFxTesting/build/assets/RunnerTemplate.Windows.txt:
```diff
@@ -1,22 +1,81 @@
+goto argparser_start
+:argparser_end
+if "%argparser_runtime_path_specification_count%" == "0"  ( goto usage )
+if "%argparser_runtime_path_specification_count%" == "1"  (
```
Do we want all this logging?

hoyosjs commented on src/Microsoft.DotNet.CoreFxTesting/build/assets/RunnerTemplate.Windows.txt:
```diff
@@ -40,7 +99,7 @@ echo ----- end %TIME% ----- exit code %ERRORLEVEL% -----------------------------
 :: ========================= END Test Execution =================================
 
-:ShowUsage
+:usage
```
Can you please update the usage messages here and in the sh:
```
echo Usage:
echo.
echo This script now uses only named arguments as follows:
echo.
echo %0 -r <runtime-dir> [-d <dotnet-host-dir>] [-g <global-tool-dir>] [-x <xunit-rsp-path>]
echo.
echo Parameters:
echo runtime-dir:            (Mandatory) Root dir containing the full test-runtime and host necessary for test execution.
echo dotnet-host-dir:     (Optional) Root dir containing the dotnet host for hosting shims and tools.
echo global-tool-dir:      (Optional) Root dir containing global tools needed.
echo xunit-rsp-path:       (Optional) File path containing additional arguments for the XUnit runner.
EXIT /B -1
```

hoyosjs replied:
Also, just as ~dp0 gets broken, %0 does too. Say `%EXECUTION_DIR%/RunTests.cmd` maybe? Not the cleanest, but this is definitely wrong (prints the last argument)

hoyosjs commented:
Fixes #2287

hoyosjs commented on src/Microsoft.DotNet.CoreFxTesting/build/assets/RunnerTemplate.Unix.txt:
```diff
@@ -1,16 +1,87 @@
+    -runtime-path|-r)
+      if [[ "$argparser_runtime_path_specified" == '1' ]] ; then
+        usage
+        exit 1
```
can you hoist the exit to the usage function and rename to `exit_with_usage`

hoyosjs commented:
Not suggesting you have to fix it, other issue is passing in a dir ending up in `\`. Not sure here, but maybe these should be standardized.

hoyosjs commented on src/Microsoft.DotNet.CoreFxTesting/build/assets/RunnerTemplate.Windows.txt:
```diff
@@ -1,22 +1,76 @@
+    if defined argparser_runtime_path_specified ( goto usage )
+    if "%~1" == "" ( goto usage )
+    set "argparser_runtime_path_specified=1"
+    set "argparser_runtime_path=%~1"
```
Nit: We know these are files, can you change to `%~f1` to make it a full path? Easier to diagnose.

hoyosjs commented:
Actually, @ViktorHofer can tell us if this is right much more easily. I thing it would be enough to add:
```
<RunArguments Condition="'$(TargetOS)' == 'Windows_NT' AND '$(ArchiveTest)' == 'true'">$(RunArguments) %XUNIT_RSP%</RunArguments>
<RunArguments Condition="'$(TargetOS)' != 'Windows_NT' AND '$(ArchiveTest)' == 'true'">$(RunArguments) $XUNIT_RSP</RunArguments>
```
inside the General XUnit Options *PropertyGroup* in `Test.targets`. I've tested this approach and it works, but I'm not sure if they have any preference over not doing this. This couples the variable names in the script to the targets, but it would expand to blank so it doesn't sound too bad.

Also, can you please verify if this all looks good? We'd like to get CoreFX testing rolling on CoreCLR.

ViktorHofer commented:
Sorry for keeping you waited here. I had some more important things to finish in corefx (fixing official builds, mono bringup). I'm now looking through the changes. I believe the scripts can still be vastly simplified. I'll give it a try.

ViktorHofer commented on src/Microsoft.DotNet.CoreFxTesting/build/core/test/Test.targets:
```diff
@@ -47,6 +47,10 @@
     <RunArguments Condition="'$(XUnitOptions)' != ''">$(RunArguments) $(XUnitOptions)</RunArguments>
 
+    <!--RunTests.cmd/RunTests.sh caller can pass a rsp file to xunit -->
+    <RunArguments Condition="'$(TargetOS)' == 'Windows_NT'">$(RunArguments) %XUNIT_RSP%</RunArguments>
```
`RunArguments` is xunit agnostic, it shouldn't be tied to it. I propose something more generic like `rsp-file` which can be used in more scenarios.

hoyosjs replied:
Well, this one is tied to our version of xunit. We'd need specific knowledge of the tools that runtests will pass it too.

ViktorHofer replied:
Sorry I wasn't clear. I meant we should change the argument name in the script. I will do that.

mikem8361 commented:
I assume this supersedes my PR #2281.  Should I close my PR?

hoyosjs commented:
@mikem yes, it does.

ViktorHofer commented:
> Should I close my PR?

Yes I'm currently working on finish this.

cshung commented:
> > Should I close my PR?
> 
> Yes I'm currently working on finish this.

Thanks @ViktorHofer for stepping up and taking over this. For your information, you can check out [my fork](https://github.com/cshung/arcade/tree/dev/andrewau/named-argument-workspace) to see more information about this pull request. 

FYI @mikem8361

ViktorHofer commented:
Updated the scripts and will test them now. @cshung @hoyosjs @mikem8361 PTAL

ViktorHofer commented:
We don't want to support both `-` and `--` for long names. `-` should only be used for aliases and `--` for the full argument name. This is different in arcade and corefx as we use powershell there which doesn't work well with the `--` syntax. In this case we don't face that problem and can therefore simplify the syntax and make it POSIX compatible.

safern commented on src/Microsoft.DotNet.CoreFxTesting/build/assets/RunnerTemplate.Windows.txt:
```diff
@@ -37,15 +83,16 @@ pushd %EXECUTION_DIR%
 
-:ShowUsage
-echo.
+:usage
```
-h|--help option missing?

safern replied:
I mean, in the usage output.

hoyosjs commented on src/Microsoft.DotNet.CoreFxTesting/build/assets/RunnerTemplate.Unix.txt:
```diff
@@ -1,91 +1,61 @@
-        exit_with_usage
-      fi
-      argparser_runtime_path=$1
+    --help|-h)
```
This changes one of the behaviors that was consciously added: if a parameter was specified twice, it was treated as an error. Sure, this was a nice behavior although I'm not married to it.

ViktorHofer replied:
Right. I'm trading readability and efficiency for this nit feature. We don't have such checks in the shell scripts in arcade or corefx either.

hoyosjs replied:
Readability does sound like it trumps anything else in this scenario, even if they are generated. All good then.

ViktorHofer replied:
Yeah my main concern with the previous appraoch was the amount of code that was added to the scripts and the decrease in readability. That's why I wanted to clean the scripts up before merging even though they were mostly working fine.

hoyosjs commented:
Other than the behavioral change, all looks good. I didn't suggest the change in Core.targets because I don't know what other testing frameworks you use other than the private XUnit runner that supports an RSP file. I guess if you'll use something msbuild-SDK based tool then it makes sense. Thanks for all the help to enable this scenario!

ViktorHofer commented:
> Thanks for all the help to enable this scenario!

Thank you and @cshung for all the help here! I'm good at refactoring and cleaning stuff up but it would probably have costed me far too much time to implement this.

The Windows tests were successful. Now testing the shell script on Ubuntu.

ViktorHofer commented:
Fixed a bug. All scenarios validated (also rsp file), works fine on both Unix and Windows.

ViktorHofer commented:
cc @filipnavara with this change you can now specify a rsp file which contains additional arguments for the test invocation. Also you don't need to specify all parameters anymore, only the runtime-path is still required. That should help Mono with bootstrapping the corefx tests.

ViktorHofer commented:
cc @marek-safar
